### ISSUE TITLE
Local map saving only saves whole blocks from the server

### ISSUE THREAD

**Issue by anon55555** [minetest/minetest#10394]
##### Minetest version

```
5.3.0
```

##### Summary
Local map saving only saves whole blocks sent from the server, but not blocks updated by `TOCLIENT_NODEMETA_CHANGED`, `TOCLIENT_ADDNODE` or `TOCLIENT_REMOVENODE`.

##### Steps to reproduce
Enable local map saving, put an item in a chest then log out. It won't be updated in the local map save.

**Comment by sfan5** [comment]
This was designed like this for simplicity but could be fixed with some extra work.

**Comment by sorcerykid** [comment]
I'm curious how something like this would be fixed. I imagine that the only feasible solution is to maintain a secondary copy of each mapblock in memory for storing the approved/reported changes from the server to be committed to the database at regular intervals (which seems like overkill). The other alternative is to commit every approved/reported change to the database immediately (which seems rather inefficient).

**Comment by rubenwardy** [comment]
The client already stores a copy of every mapblock in range - it's how you're able to see nearby blocks. As for distant changes, clients won't be sent the change packets

**Comment by sorcerykid** [comment]
> The client already stores a copy of every mapblock in range

Yes, but the local player is able to dig and place nodes without approval from the server, meaning that copy isn't guaranteed to always mirror the server.

**Comment by anon55555** [comment]
The client should remember place predicts so it can revert them if the server doesn't confirm them after a timeout, this would fix https://github.com/minetest/minetest/issues/10410.

**Comment by anon55555** [comment]
This could be done by storing a copy of blocks that contain place predictions and deleting the copy after a timeout.
